compiler/src/iree/compiler/Dialect/Util/Transforms/VerifyInitializationOrder.cpp

```diff
@@ -0,0 +1,407 @@
+            auto globalName = storeOp.getGlobalName();
+            auto &global = globalTable.lookup(globalName);
+            if (global.op && global.op.getGlobalInitialValue()) {
+              // Global has an initial value. Check if it comes after this
+              // initializer.
+              size_t globalOrdinal = opOrdinalMap[global.op];
+              if (globalOrdinal > initOrdinal) {
+                emitDiagnostic(
+                    ERROR, storeOp->getLoc(),
+                    "global '@" + globalName +
+                        "' with initial value at position " +
+                        Twine(globalOrdinal) +
+                        " is modified by earlier initializer at position " +
+                        Twine(initOrdinal));
+                return WalkResult::interrupt();
+              }
+            }
+            return WalkResult::advance();
+          });
+      if (result.wasInterrupted()) {
+        return failure();
+      }
+    }
+    return success();
+  }
+
+  // Computes which functions are only reachable from initializers.
+  InitializerReachability
+  computeInitializerReachability(mlir::ModuleOp moduleOp) {
+    InitializerReachability result;
+
+    // Build the call graph.
+    CallGraph callGraph(moduleOp);
+
+    // Find all externally reachable functions.
+    SetVector<CallGraphNode *> externalWorklist;
+    for (CallGraphNode *node : callGraph) {
+      if (node->isExternal()) {
+        continue;
+      }
+      auto *callableOp = node->getCallableRegion()->getParentOp();
+      if (isa<IREE::Util::InitializerOpInterface>(callableOp)) {
+        // Initializers are not externally reachable.
+        continue;
+      }
+      if (auto funcOp = dyn_cast<FunctionOpInterface>(callableOp)) {
+        if (funcOp.isPublic()) {
+          // Public function - externally reachable.
+          result.externallyReachableFunctions.insert(callableOp);
+          externalWorklist.insert(node);
+        }
+      }
+    }
+
+    // Transitively mark all functions called from external entry points.
+    while (!externalWorklist.empty()) {
+      auto *node = externalWorklist.pop_back_val();
+      for (auto &edge : *node) {
+        auto *targetNode = edge.getTarget();
+        if (!targetNode->isExternal()) {
+          auto *targetOp = targetNode->getCallableRegion()->getParentOp();
+          if (result.externallyReachableFunctions.insert(targetOp).second) {
+            externalWorklist.insert(targetNode);
+          }
+        }
+      }
+    }
+
+    // Find functions only called from initializers.
+    SetVector<CallGraphNode *> initializerWorklist;
+    for (CallGraphNode *node : callGraph) {
+      if (node->isExternal()) {
+        continue;
+      }
+      auto *callableOp = node->getCallableRegion()->getParentOp();
+      if (isa<IREE::Util::InitializerOpInterface>(callableOp)) {
+        initializerWorklist.insert(node);
+      }
+    }
```

Comment:
Why don't we combine this with `externalWorklist` loop, i.e., move the insert to line 258?